### ISSUE TITLE
How to properly issue refunds?

### ISSUE THREAD

**Issue by kramaswamy7** [intuit/QuickBooks-V3-DotNET-SDK#152]
Hi Nimisha,

Trying to understand how to correctly model a refund using the .NET SDK.

Roughly speaking, here's how we are handling payments (`transaction` is an internal entity with info about the transaction):

```
Payment payment = new Payment();

payment.TxnDate = transaction.TxnDate;
payment.TxnDateSpecified = true;

payment.TotalAmt = transaction.TotalAmt;
payment.TotalAmtSpecified = true;
payment.PrivateNote = transaction.PrivateNote;

payment.PaymentRefNum = transaction.PaymentRefNum;
payment.CustomerRef = transaction.CustomerRef;
payment.PaymentMethodRef = new ReferenceType { name = transaction.PaymentMethod.Name, Value = transaction.PaymentMethod.Id };
payment.DepositToAccountRef = ToReferenceType(GetAccountByName("Undeposited Funds"));

Batch batch = CreateNewBatch();

batch.Add(payment, 0, OperationEnum.create);

batch.Execute();
```

I attempted to create a refund similarly:

```
RefundReceipt refund = new RefundReceipt();

refund.TxnDate = transaction.TxnDate;
refund.TxnDateSpecified = true;

refund.TotalAmt = transaction.TotalAmt * -1;
refund.TotalAmtSpecified = true;
refund.PrivateNote = transaction.PrivateNote;

refund.PaymentRefNum = transaction.PaymentRefNum;
refund.CustomerRef = transaction.CustomerRef;
refund.PaymentMethodRef = new ReferenceType { name = transaction.PaymentMethod.Name, Value = transaction.PaymentMethod.Id };
refund.DepositToAccountRef = ToReferenceType(GetAccountByName("Undeposited Funds"));

Batch batch = CreateNewBatch();

batch.Add(payment, 0, OperationEnum.create);

batch.Execute();
```

However, this threw an error suggesting that Line was missing. Reading the documentation on RefundReceipt located at https://developer.intuit.com/app/developer/qbo/docs/api/accounting/all-entities/refundreceipt, I see that it does indeed indicate the Line array is mandatory. I tried adding it in as so:

```
List<Line> lines = new List<Line>();

lines.Add(new Line
{
    DetailType = LineDetailTypeEnum.SalesItemLineDetail,
    DetailTypeSpecified = true,
    Amount = transaction.TotalAmt * -1
});

refund.Line = lines.ToArray();
```

However, Line indicates `SalesLineItem` or `GroupLineItem` as mandatory properties. Line, though, is of type `Line[]`, and `Line` does not include a `SalesItemLineDetail` property, even though the documentation suggests it is required.

I could probably hack something together using reflection, but that doesn't seem like the correct approach. Also, even if I were to get it working correctly, `SalesLineItem` has several properties that I'm unsure of how to set.

Any insights would be appreciated. Thanks!

**Comment by nimisha84** [comment]
Here are some examples that can help you-
https://gist.github.com/IntuitDeveloperRelations/6500373

Check CreateASTInvoice or CreateInvoice here- https://github.com/IntuitDeveloper/SampleApp-CRUD_.Net_Oauth2/blob/master/SampleApp_CRUD_.Net/SampleApp_CRUD_.Net/Helper/QBOHelper.cs

**Comment by kramaswamy7** [comment]
Hi Nimisha,

That does get me closer, thanks. Still unsure about one thing though -

In the documentation it suggests for ItemRef "Query the Item name list resource to determine the appropriate Item object for this reference. Use Item.Id and Item.Name from that object for ItemRef.value and ItemRef.name, respectively." It then says "Applicable to invoice objects, only, and when linktxn specifies a ReimburseCharge. When Item.Id is set to 1, ItemAccountRef refers to reimburse expense account Id."

Basically - I'm not sure what should be specified for `Item.Id` and `Item.Name`, nor for `ItemAccountRef`. Is `Item.Id` and `Item.Name` supposed to refer to the original transaction that is being reimbursed?

**Comment by nimisha84** [comment]
Just query an existing item and use it's id and name(optional) in your ref fields.
Ignore the ItemAccountRef. Thst is just a note for devs who use Reimburse charge liknked txn within the invoice.
